src/backend/distributed/connection/placement_connection.c

```diff
@@ -229,272 +222,373 @@ GetPlacementConnection(uint32 flags, ShardPlacement *placement, const char *user
 	 */
-	if (placementEntry->modifyingConnection)
+	foreach(placementAccessCell, placementAccessList)
 	{
-		ConnectionReference *connectionReference = placementEntry->modifyingConnection;
+		ShardPlacementAccess *placementAccess =
+			(ShardPlacementAccess *) lfirst(placementAccessCell);
+		ShardPlacement *placement = placementAccess->placement;
+		ShardPlacementAccessType accessType = placementAccess->accessType;
+
+		ConnectionPlacementHashEntry *placementEntry = NULL;
+		ColocatedPlacementsHashEntry *colocatedEntry = NULL;
+		ConnectionReference *placementConnection = NULL;
 
-		if (CanUseExistingConnection(flags, userName, connectionReference))
+		if (placement->shardId == INVALID_SHARD_ID)
 		{
-			return connectionReference;
+			/*
+			 * When a SELECT prunes down to 0 shard, we use a dummy placement.
+			 * In that case, we can fall back to the default connection.
+			 *
+			 * FIXME: this can be removed if we evaluate empty SELECTs locally.
+			 */
+			if (lastPlacement == NULL)
+			{
+				lastPlacement = placement;
+			}
+			continue;
 		}
 
-		if (connectionReference->hadDDL)
+		placementEntry = FindOrCreatePlacementEntry(placement);
+		colocatedEntry = placementEntry->colocatedEntry;
+		placementConnection = placementEntry->primaryConnection;
+
+		if (placementConnection->connection == NULL)
 		{
-			/* would deadlock otherwise */
+			/* no connection has been chosen for the placement */
+		}
+		else if (accessType == PLACEMENT_ACCESS_DDL &&
+				 placementEntry->hasSecondaryConnections)
+		{
+			/*
+			 * If a placement has been read over multiple connections (typically as
+			 * a result of a reference table join) then a DDL command on the placement
+			 * would create a self-deadlock.
+			 */
 			ereport(ERROR,
 					(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
-					 errmsg("cannot establish new placement connection when DDL has been "
-							"executed on existing placement connection")));
+					 errmsg("cannot perform DDL on a placement which has been read over "
+							"multiple connections")));
 		}
-		else if (connectionReference->hadDML)
+		else if (accessType == PLACEMENT_ACCESS_DDL && colocatedEntry != NULL &&
+				 colocatedEntry->hasSecondaryConnections)
 		{
-			/* we'd not see the other connection's contents */
+			/*
+			 * If a placement has been read over multiple (uncommitted) connections
+			 * then a DDL command on a co-located placement may create a self-deadlock
+			 * if there exist some relationship between the co-located placements
+			 * (e.g. foreign key, partitioning).
+			 */
 			ereport(ERROR,
 					(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
-					 errmsg("cannot establish new placement connection when DML has been "
-							"executed on existing placement connection")));
+					 errmsg("cannot perform DDL on a placement if a co-located placement "
+							"has been read over multiple connections")));
 		}
-		else
+		else if (foundModifyingConnection)
```

Comment:
There are several improvements we can make to increase the set of multi-statement transactions that are supported, which avoids these situations:

#1482 - enable single DML / COPY + DDL/COPY (should aim to fix this in 7.0)
#1273 - enable multiple DML + DDL
#1484 - enable multiple DML + COPY

We cannot support your last example. 

I don't see an easy way to error out in sometimes-unsupported scenarios. I think we'd need to start keeping track of things at a table level. If non-co-located tables have been modified over multiple connections then we should error out if a query reads from both of those tables.